cpp/src/arrow/compute/kernels/take-internal.h

```diff
@@ -0,0 +1,573 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <algorithm>
+#include <limits>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "arrow/builder.h"
+#include "arrow/compute/context.h"
+#include "arrow/util/bit-util.h"
+#include "arrow/util/checked_cast.h"
+#include "arrow/util/logging.h"
+#include "arrow/util/stl.h"
+#include "arrow/visitor_inline.h"
+
+namespace arrow {
+namespace compute {
+
+using internal::checked_cast;
+using internal::checked_pointer_cast;
+
+template <typename Builder, typename Scalar>
+static Status UnsafeAppend(Builder* builder, Scalar&& value) {
+  builder->UnsafeAppend(std::forward<Scalar>(value));
+  return Status::OK();
+}
+
+static Status UnsafeAppend(BinaryBuilder* builder, util::string_view value) {
+  RETURN_NOT_OK(builder->ReserveData(static_cast<int64_t>(value.size())));
```

Comment:
That's not a problem. I'll write a benchmark so we can watch the difference.

Reply:
@pitrou  as it turns out, it's actually a little faster to ReserveData for each string:

```
# ReserveData on append
---------------------------------------------------------------------------------------------
Benchmark                                      Time           CPU Iterations UserCounters...
---------------------------------------------------------------------------------------------
FilterString/1048576/1/min_time:1.000    1586715 ns    1586705 ns        884 null_percent=1 size=1048.58k   630.237MB/s
FilterString/8388608/1/min_time:1.000   23555110 ns   23555006 ns         72 null_percent=1 size=8.38861M   339.631MB/s

# separate ReserveData pass
---------------------------------------------------------------------------------------------
Benchmark                                      Time           CPU Iterations UserCounters...
---------------------------------------------------------------------------------------------
FilterString/1048576/1/min_time:1.000    1656093 ns    1654443 ns        841 null_percent=1 size=1048.58k   604.433MB/s
FilterString/8388608/1/min_time:1.000   19849301 ns   19809235 ns         71 null_percent=1 size=8.38861M   403.852MB/s
```

Would you like me to push this change so you can inspect it?

Reply:
It's faster on the 1MB case, but slower on the 8MB case. I expect it to become even slower on larger data. Can you try with e.g. 32MB?

Reply:
(what is the size of your L3 cache?)

Reply:
Update: we discussed and investigated it a bit with @bkietz . It turns out that the dominating contributor to Filter performance is probably reading the selection bitmap, so iterating twice on the input instead of once has a negative impact.